File: js/web/lib/wasm/jsep/webgpu/ops/scatter-nd.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { DataType } from '../../../wasm-common';
import { TensorView } from '../../tensor-view';
import { ShapeUtil } from '../../util';
import { AttributeWithCacheKey, createAttributeWithCacheKey } from '../attribute-with-cache-key';
import { ComputeContext, ProgramInfo, ProgramUniform } from '../types';

import {
  atomicOutputVariable,
  createTensorShapeVariables,
  inputVariable,
  outputVariable,
  ShaderHelper,
} from './common';

export interface ScatterNDAttributes extends AttributeWithCacheKey {
  reduction: string;
}

type ReductionType = 'i32' | 'u32' | 'f32';

const atomicReductionSnippet = (reduction: string, ptr: string, v: string, type: ReductionType) => {
  if (reduction !== 'none' && type !== 'i32' && type !== 'u32' && type !== 'f32') {
    throw new Error(`Input ${type} is not supported with reduction ${reduction}.`);
  }

  const floatStart = `{
                var oldValue = 0;
                loop {
                  let newValueF32 =`;
  const floatEnd = `;
                  let newValue = bitcast<i32>(newValueF32);
                  let res = atomicCompareExchangeWeak(&${ptr}, oldValue, newValue);
                  if res.exchanged {
                    break;
                  }
                  oldValue = res.old_value;
                }
              }`;

  switch (reduction) {
    case 'none':
      return `${ptr}=${v};`;
    case 'add':
      if (type === 'i32' || type === 'u32') {
        return `atomicAdd(&${ptr}, bitcast<${type}>(${v}));`;
      } else {
        // atomicAdd only supports uint/int type. For float, we use
        // atomicCompareExchangeWeak to simulate.
        return `
              ${floatStart}bitcast<${type}>(oldValue) + (${v})${floatEnd}`;
      }
    case 'max':
      if (type === 'i32' || type === 'u32') {
        return `atomicMax(&${ptr}, bitcast<${type}>(${v}));`;
      } else {
        // atomicMax only supports uint/int type. For float, we use
        // atomicCompareExchangeWeak to simulate.
        return `
                ${floatStart}max(bitcast<f32>(oldValue), (${v}))${floatEnd}`;
      }
    case 'min':
      if (type === 'i32' || type === 'u32') {
        return `atomicMin(&${ptr}, bitcast<${type}>(${v}));`;
      } else {
        // atomicMin only supports uint/int type. For float, we use
        // atomicCompareExchangeWeak to simulate.
        return `${floatStart}min(bitcast<${type}>(oldValue), (${v}))${floatEnd}`;
      }
    case 'mul':
      // atomicMul is not supported, we use atomicCompareExchangeWeak to simulate.
      return `${floatStart}(bitcast<${type}>(oldValue) * (${v}))${floatEnd}`;

    default:
      throw new Error(`Reduction ${reduction} is not supported.`);
  }
};

const createScatterNDProgramInfo = (inputs: readonly TensorView[], attributes: ScatterNDAttributes): ProgramInfo => {
  const inputShape = inputs[0].dims;
  const indicesShape = inputs[1].dims;
  const outputShape = inputShape;
  // TODO: support bool with components 4.
  const components = 1;
  const outputSize = Math.ceil(ShapeUtil.size(indicesShape) / components);
  const lastIndexDimension = indicesShape[indicesShape.length - 1];
  const numUpdatesElements = ShapeUtil.sizeFromDimension(inputShape, lastIndexDimension);

  const programUniforms: ProgramUniform[] = [
    { type: DataType.uint32, data: outputSize },
    { type: DataType.uint32, data: lastIndexDimension },
    { type: DataType.uint32, data: numUpdatesElements },
    ...createTensorShapeVariables(inputs[1].dims, inputs[2].dims, outputShape),
  ];

  const getShaderSource = (shaderHelper: ShaderHelper) => {
    const indices = inputVariable('indices', inputs[1].dataType, inputs[1].dims.length);
    const updates = inputVariable('updates', inputs[2].dataType, inputs[2].dims.length, components);
    const output =
      attributes.reduction !== 'none' && attributes.reduction !== ''
        ? atomicOutputVariable('output', inputs[0].dataType, outputShape.length)
        : outputVariable('output', inputs[0].dataType, outputShape.length, components);

    return `
      ${shaderHelper
        .registerUniform('output_size', 'u32')
        .registerUniform('last_index_dimension', 'u32')
        .registerUniform('num_updates_elements', 'u32')
        .declareVariables(indices, updates, output)}
      ${shaderHelper.mainStart()}
        ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes('uniforms.output_size')}
  var data_offset = 0u;
  let indices_start = uniforms.last_index_dimension * global_idx;
  let indices_end = indices_start + uniforms.last_index_dimension;
  for (var i = indices_start; i < indices_end; i++) {
    var index = i32(indices[i].x);
    ${
      inputs[0].dims.length === 1
        ? `
    let element_count_dim = uniforms.output_strides;
    let dim_value = uniforms.output_shape;`
        : `
    let element_count_dim = uniforms.output_strides[i - indices_start];
    let dim_value = uniforms.output_shape[i - indices_start + uniforms.last_index_dimension];`
    }
    if (index >= 0) {
      if (index >= i32(dim_value)) {
        index = i32(dim_value - 1);
      }
    } else {
      if (index < -i32(dim_value)) {
        index = 0;
      } else {
        index += i32(dim_value);
      }
    }
    data_offset += u32((u32(index) * element_count_dim));
  }

  for (var i = 0u; i < uniforms.num_updates_elements; i++) {
    let value = updates[uniforms.num_updates_elements * global_idx + i];
    ${atomicReductionSnippet(
      attributes.reduction,
      'output[data_offset + i]',
      'value',
      output.type.value as ReductionType,
    )}
  }

      }`;
  };
  return {
    name: 'ScatterND',
    shaderCache: {
      hint: `${attributes.cacheKey}_${attributes.reduction}`,
      inputDependencies: ['rank', 'rank'],
    },
    getRunData: () => ({
      outputs: [{ dims: outputShape, dataType: inputs[0].dataType }],
      dispatchGroup: { x: Math.ceil(outputSize / 64 /* workgroup size */) },
      programUniforms,
    }),
    getShaderSource,
  };
};

export const parseScatterNDAttributes = (attributes: Record<string, unknown>): ScatterNDAttributes =>
  createAttributeWithCacheKey({ reduction: attributes.reduction as string });

export const scatterND = (context: ComputeContext, attributes: ScatterNDAttributes): void => {
  context.compute(createScatterNDProgramInfo(context.inputs, attributes), {
    inputs: [context.inputs[1], context.inputs[2]],
    outputs: [],
  });
};

File: onnxruntime/core/providers/webgpu/tensor/scatter_nd.cc
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <string>
#include <iostream>
#include "core/providers/webgpu/webgpu_kernel.h"
#include "core/providers/webgpu/webgpu_supported_types.h"
#include "core/providers/webgpu/shader_helper.h"
#include "core/providers/webgpu/shader_variable.h"
#include "scatter_nd.h"

namespace onnxruntime {
namespace webgpu {

Status ScatterNDProgram::GenerateShaderCode(ShaderHelper& shader) const {
  shader.AddInput("indices", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  shader.AddInput("updates", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const auto& output = shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseShapeAndStride);
  const auto output_rank = static_cast<size_t>(output.Rank());
  auto atomic_reduction_snippet = [](ScatterNDReduction reduction, const std::string& ptr, const std::string& value, const std::string& data_type) -> std ::string {
    std::ostringstream ss;
    bool is_32_bit_integer = data_type == "i32" || data_type == "u32";
    bool is_unsigned_integer = data_type == "u32";
    std::ostringstream ss_float_start;
    ss_float_start << "    {\n"
                   << "      var oldValue = 0" << (is_unsigned_integer ? "u" : "") << ";\n"
                   << "      loop {\n"
                   << "        let newValueF32 = ";
    std::ostringstream ss_float_end;
    ss_float_end << ";\n"
                 << "          let newValue = bitcast<" << (is_unsigned_integer ? "u32" : "i32") << ">(newValueF32);\n"
                 << "          let res = atomicCompareExchangeWeak(&" << ptr << ", oldValue, newValue);\n"
                 << "          if res.exchanged {\n"
                 << "            break;\n"
                 << "          }\n"
                 << "          oldValue = res.old_value;\n"
                 << "        }\n"
                 << "      }\n";
    switch (reduction) {
      case ScatterNDReduction::None:
        ss << "    " << ptr << " = " << value << ";\n";
        break;
      case ScatterNDReduction::Add:
        if (is_32_bit_integer) {
          ss << "  atomicAdd(&" << ptr << ", bitcast<" << data_type << ">(" << value << "));\n";
        } else {
          // atomicAdd only supports uint/int type. For float, we use
          // atomicCompareExchangeWeak to simulate.
          ss << ss_float_start.str() << "bitcast<" << data_type << ">(oldValue) + (" << value << ")" << ss_float_end.str()
             << "\n";
        }
        break;
      case ScatterNDReduction::Max:
        if (is_32_bit_integer) {
          ss << "  atomicMax(&" << ptr << ", bitcast<" << data_type << ">(" << value << "));\n";
        } else {
          // atomicMax only supports uint/int type. For float, we use
          // atomicCompareExchangeWeak to simulate.
          ss << ss_float_start.str() << "max(bitcast<" << data_type << ">(oldValue), (" << value << "))" << ss_float_end.str();
        }
        break;
      case ScatterNDReduction::Min:
        if (is_32_bit_integer) {
          ss << "  atomicMin(&" << ptr << ", bitcast<" << data_type << ">(" << value << "));\n";
        } else {
          // atomicMin only supports uint/int type. For float, we use
          // atomicCompareExchangeWeak to simulate.
          ss << ss_float_start.str() << "min(bitcast<" << data_type << ">(oldValue), (" << value << "))" << ss_float_end.str();
        }
        break;
      case ScatterNDReduction::Mul:
        // atomicMul is not supported, we use atomicCompareExchangeWeak to simulate.
        ss << ss_float_start.str() << "(bitcast<" << data_type << ">(oldValue) * (" << value << "))" << ss_float_end.str();
        break;
      default:
        ORT_THROW("Unsupported reduction type: ", static_cast<int>(reduction));
        // The controlflow should never reach here.
    }
    return ss.str();
  };

  auto calc_data_offset_snippet = [](size_t output_rank) -> std::string {
    std::ostringstream ss;
    if (output_rank < 2) {
      ss << "    let element_count_dim = 1u;\n";
    } else {
      ss << "    let element_count_dim = select(" << GetElementAt("uniforms.output_stride", "i - indices_start", output_rank - 1) << ", 1u, i - indices_start == " << (output_rank - 1) << ");\n";
    }
    ss << "    let dim_value = " << GetElementAt("uniforms.output_shape", "i - indices_start", output_rank) << ";\n"
       << "    if (index >= 0) {\n"
       << "      if (index >= i32(dim_value)) {\n"
       << "        index = i32(dim_value - 1);\n"
       << "      }\n"
       << "    } else {\n"
       << "      if (index < -i32(dim_value)) {\n"
       << "        index = 0;\n"
       << "      } else {\n"
       << "        index += i32(dim_value);\n"
       << "      }\n"
       << "    }\n"
       << "    data_offset += u32((u32(index) * element_count_dim));\n";
    return ss.str();
  };

  auto update_elements_snippet = [atomic_reduction_snippet](ScatterNDReduction reduction, const std::string& data_type) -> std::string {
    std::ostringstream ss;
    ss << "  for (var i = 0u; i < uniforms.num_updates_elements; i++) {\n"
       << "    let value = updates[uniforms.num_updates_elements * global_idx + i];\n"
       << atomic_reduction_snippet(reduction, "output[data_offset + i]", "value", data_type) << "\n"
       << "  }\n";
    return ss.str();
  };
  std::string data_type_str;
  bool reducible = false;
  if (data_type_ == DataTypeImpl::GetType<int32_t>()) {
    reducible = true;
    data_type_str = "i32";
  } else if (data_type_ == DataTypeImpl::GetType<uint32_t>()) {
    reducible = true;
    data_type_str = "u32";
  } else if (data_type_ == DataTypeImpl::GetType<float>()) {
    reducible = true;
    data_type_str = "f32";
  } else {
    // Default value.
    data_type_str = "output_element_t";
  }
  if (reduction_ != ScatterNDReduction::None && !reducible) {
    ORT_THROW("ScatterND: Reduction is not supported for data type ", data_type_str);
  }
  shader.MainFunctionBody() << shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size")
                            << "  var data_offset = 0u;\n"
                            << "  var indices_start = uniforms.last_index_dimension * global_idx;\n"
                            << "  var indices_end = indices_start + uniforms.last_index_dimension;\n"
                            << "  for (var i = indices_start; i < indices_end; i++) {\n"
                            << "    var index = i32(indices[i].x);\n"
                            << calc_data_offset_snippet(output_rank)
                            << "  }\n"
                            << update_elements_snippet(reduction_, data_type_str);
  return Status::OK();
}

Status ScatterND::ComputeInternal(ComputeContext& context) const {
  const Tensor* input = context.Input<Tensor>(0);
  const auto* indices = context.Input<Tensor>(1);
  const auto* updates = context.Input<Tensor>(2);
  const auto& input_shape = input->Shape();
  const auto& indices_shape = indices->Shape();
  auto indices_rank = indices_shape.NumDimensions();
  auto last_index_dimension = static_cast<uint32_t>(indices_shape[indices_rank - 1]);
  auto num_updates_elements = static_cast<uint32_t>(input_shape.SizeFromDimension(last_index_dimension));
  // TODO: support bool with components 4.
  const size_t components = 1;
  auto output_size = static_cast<uint32_t>((indices_shape.SizeToDimension(indices_rank - 1) + components - 1) / components);
  auto* output = context.Output(0, input_shape);
  if (output_size == 0) {
    // If the output tensor is empty, we can return early.
    return Status::OK();
  }
  MLDataType data_type = input->DataType();
  const void* source = input->DataRaw();
  void* target = output->MutableDataRaw();
  // If source and target pointers are not equal (non-inplace operation), we need to copy the data.
  if (target != source) {
    ORT_RETURN_IF_ERROR(Info().GetDataTransferManager().CopyTensor(*input, *output));
  }
  ScatterNDProgram program(reduction_, data_type);
  program
      .CacheHint(static_cast<uint32_t>(reduction_))
      .AddInputs({{indices, ProgramTensorMetadataDependency::TypeAndRank},
                  {updates, ProgramTensorMetadataDependency::TypeAndRank}})
      .SetDispatchGroupSize((output_size + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE)
      .AddUniformVariables({output_size, last_index_dimension, num_updates_elements});
  if (reduction_ != ScatterNDReduction::None && (data_type == DataTypeImpl::GetType<float>() || data_type == DataTypeImpl::GetType<int32_t>() ||
                                                 data_type == DataTypeImpl::GetType<uint32_t>())) {
    program.AddOutput({output, ProgramTensorMetadataDependency::TypeAndRank, ProgramOutput::Atomic});
  } else {
    program.AddOutput({output, ProgramTensorMetadataDependency::TypeAndRank});
  }
  return context.RunProgram(program);
}

ONNX_OPERATOR_KERNEL_EX(
    ScatterND,
    kOnnxDomain,
    18,
    kWebGpuExecutionProvider,
    (*KernelDefBuilder::Create())
        .TypeConstraint("T", WebGpuSupportedNumberTypes())
        .MayInplace(0, 0),
    ScatterND);

ONNX_OPERATOR_VERSIONED_KERNEL_EX(
    ScatterND,
    kOnnxDomain,
    16,
    17,
    kWebGpuExecutionProvider,
    (*KernelDefBuilder::Create())
        .TypeConstraint("T", WebGpuSupportedNumberTypes())
        .MayInplace(0, 0),
    ScatterND);

ONNX_OPERATOR_VERSIONED_KERNEL_EX(
    ScatterND,
    kOnnxDomain,
    13,
    15,
    kWebGpuExecutionProvider,
    (*KernelDefBuilder::Create())
        .TypeConstraint("T", WebGpuSupportedNumberTypes())
        .MayInplace(0, 0),
    ScatterND);

ONNX_OPERATOR_VERSIONED_KERNEL_EX(
    ScatterND,
    kOnnxDomain,
    11,
    12,
    kWebGpuExecutionProvider,
    (*KernelDefBuilder::Create())
        .TypeConstraint("T", WebGpuSupportedNumberTypes())
        .MayInplace(0, 0),
    ScatterND);
}  // namespace webgpu
}  // namespace onnxruntime
